Feat: new UI components

fjebaker opened the pull request.

fjebaker committed:
chore: added pyproject.toml

Migrating to use pdm instead of poetry or pipenv.

fjebaker committed:
chore: add runner script

fjebaker committed:
feat: replaced player gathering dialog

Updated the player gathering dialog with the new UI component features.
Implemented a `TimedView` class as a helper for creating timeout views
that display the time remaining in the embed.

Needs polish.

fjebaker committed:
chore: add pyproject scripts for lint and more

fjebaker committed:
feat: updated elash to use new UI

Converted the DMs system to use the new discord UI components, and
created a `UserUniqueView` to help tailor messages to single users in a
shared channel. This is quite limited unfortunately, as I can't find in
the docs a way to send an ephemeral message in a channel without going
through the `interaction.response.send_message` method.

As such, the user is asked to click a button to get their prompt, which
is a suitable workaround I suppose.

The modal input + safeties is a little funky, so we know determine the
safeties before asking for answers so we can actually display what the
safety is. Consequently the UI components return both the message (which
may be the default) and a flag to indicate if the safety was used or
not.

Needs a lot of polish. The interaction messages can be embeds too, so we
can get pretty text formatting.

Not extensively tested.

fjebaker commented on src/interactive/selection.py:
This file needs to be renamed to reflect what it is.

fjebaker commented on src/interactive/selection.py:
```diff
@@ -0,0 +1,48 @@
+
+import discord
+
+logger = logging.Logger(__name__)
```
Logger is currently unused.

fjebaker commented on src/interactive/userunique.py:
```diff
@@ -0,0 +1,108 @@
+class PromptModal(discord.ui.Modal, title="e-bot"):
+    def __init__(self, content):
+        super().__init__()
+        self.response = discord.ui.TextInput(label=content)
```
May want to try out different text input types. The default one seems to remember your inputs (at least on discord browser) which might get a bit irritating after many games. Could be the paragraph text input is more amnesiac?

fjebaker commented on src/abstracts/egamefactory.py:
```diff
@@ -23,6 +26,41 @@ def replace_rules(content: str) -> str:
     )
 
 
+class GatherPlayersView(TimedView):
+    def __init__(self, embed):
+        super().__init__(embed, timeout=PLAYER_GATHER_TIMEOUT)
+        self.players = []
+        self.text: str = self.embed.description
+
+    async def _update_player_list(self):
+        text = (
+            self.text + "\n\nPlayers:\n" + "\n -> ".join((i.name for i in self.players))
+        )
+        await self.update_text(text)
+
+    async def interaction_check(self, interaction: discord.Interaction):
+        # don't let users with the same id interact twice
+        is_playing = any((i.id == interaction.user.id for i in self.players))
+
+        if is_playing:
+            # let the player know they have already joined
+            username = interaction.user.name
+            await interaction.response.send_message(
+                f"@{username}: you have already joined."
+            )
+            return False
+
+        return True
+
+    @discord.ui.button(label="Join", style=discord.ButtonStyle.primary)
+    async def button_callback(self, interaction: discord.Interaction, button):
+        # append user
+        self.players.append(interaction.user)
+        await self._update_player_list()
+        # keep listening for more events
+        await interaction.response.defer()
+
+
```
This should be moved to the `interactions` module.

fjebaker commented on src/econfig.py:
```diff
@@ -1,2 +1,2 @@
 PATH_EXTENSION = "."
-PLAYER_GATHER_TIMEOUT = 16
+PLAYER_GATHER_TIMEOUT = 5
```
Needs reverting!

JR-Mitchell commented on src/abstracts/egamefactory.py:
```diff
@@ -189,17 +226,12 @@ async def _players_prompt(self) -> dict:
 
-        # create reply reaction pipeline
-        ipl = InteractionPipeline(ReplyInteraction())
+        text = f"{self.game_description}\n\nReply to this message to join the game."
```
This copy needs updating

JR-Mitchell commented on pyproject.toml:
I presume this is a new type of dependencies / build running spec?

fjebaker replied:
https://github.com/pdm-project/pdm

JR-Mitchell commented on src/abstracts/egamefactory.py:
```diff
@@ -23,6 +26,41 @@ def replace_rules(content: str) -> str:
+
+    async def _update_player_list(self):
+        text = (
+            self.text + "\n\nPlayers:\n" + "\n -> ".join((i.name for i in self.players))
```
Should be:
```
self.text + "\n\nPlayers:\n-> " + "\n -> ".join((i.name for i in self.players))
```
to ensure that the first player also has `->`

JR-Mitchell commented on src/games/elash.py:
```diff
@@ -31,13 +36,12 @@ def randomize_prompts(prompts: list) -> list:
     game_description = ""
     wait_duration = 5
-    min_players = 2
+    min_players = 1
```
Needs reverting too?

fjebaker replied:
Yes

JR-Mitchell commented on src/interactive/userunique.py:
```diff
@@ -0,0 +1,108 @@
+        self.outcome = modal.response.value
+        await self.resolve(interaction.message, modal.interaction)
+
+    async def on_timeout(self):
```
What is the behaviour here if the component times out but one of the `await` statements (e.g `await modal.wait()`) is still executing?

fjebaker replied:
I believe the discord library kills any hanging coros in that event loop, calls `on_timeout`, and the yields to wherever `await thing.wait()` is?

fjebaker committed:
feat: ui fixes and poll

Added a `PollView` with the new UI components.

Updated elash to use new `PollView` component.

Intermediary results now delete themselves after some timeout.

Fixed bugs related to being able to vote and enter two results by
whitelisting which user IDs may interact and checking whether a user has
already interacted.

Fixed bugs related to getting input from modal text inputs when the
label text was >45 characters by hardcoding the label to "Answer".

Timed UI components now skip the timer when everyone who is expected to
have interacted has indeed interacted.

fjebaker committed:
fix: revert timeout for gathering and min players in elash

fjebaker committed:
fix: increase prompt time and vote tallies

Bumped the time to submit prompts from 15 seconds to 30.

Show the vote tallies in the final vote screen as the label of the embed
field.

fjebaker committed:
feat: emoji hell

Added a big list of emojis and a random emoji function. This is used to
assign a random emoji to each player, and to all of the UI buttons for a
bit more flair.

Downside of this is that we can no longer use the button decorators, as
the random emoji function would only be called during module load. Moved
all relevant buttons into the body.

Added a (useless?) helper function for context wrapping async functions
to pass the `self` argument. It seems Python already does that when
qualifying with `self.function_name`?

fjebaker committed:
fix: emoji mapping can't fail

There are some string codes that don't map to emojis, so we just keep
trying to convert until we get one. Bad strings get logged as warnings
so I can maybe one day grep through the logs and remove them.

fjebaker committed:
style: formatting

fjebaker committed:
fix: timeouts for elash rounds and safety tally

Fixed logic error when deducing which players used safeties. We now
instead just track whether a reply was a safety or not with a boolean
attached to the reply itself.

Timeout for votes and rounds increased to 1 minute, from 30 seconds.

fjebaker committed:
chore: update requirements
